### PR TITLE
first cut at proper read/write locking for KeyBundle

### PR DESCRIPTION
**Pull request by jschlyter**
KeyBundle is tricky as it updates its own keys from multiple places in the code. This PR uses [readwriterlock](https://pypi.org/project/readerwriterlock/) to put a read lock around key reads and a write lock updates.

The write lock is implemented using a decorator, whereas the read lock implemented using explicit `with`. Most key access has been moved to `keys()` in order to simplify the code.

There may be dragons - please review carefully. Extra set of eyes from @janste63 would be most useful.

### CODE REVIEW

**Review by rohe — APPROVED** [review]

File: src/cryptojwt/key_bundle.py [hunk 1]
```diff
@@ -551,32 +561,34 @@ def get(self, typ="", only_active=True):
...
-        if typ:
-            _keys = [k for k in self._keys if k.kty in _typs]
-        else:
-            _keys = self._keys
+        with self._lock_reader:
+            if typ:
+                _typs = [typ.lower(), typ.upper()]
+                _keys = [k for k in self._keys if k.kty in _typs]
+            else:
+                _keys = self._keys
```
Comment (janste63) [inline src/cryptojwt/key_bundle.py]: Need a shallow copy here

File: src/cryptojwt/key_bundle.py [hunk 2]
```diff
@@ -551,32 +561,34 @@ def get(self, typ="", only_active=True):
...
 
         :return: List of all keys
         """
-        self._uptodate()
-
-        return self._keys
+        if update:
+            self._uptodate()
+        with self._lock_reader:
+            return self._keys
```
Comment (janste63) [inline src/cryptojwt/key_bundle.py]: Need a shallow copy here. Optimization, save the copy and remove copy every time keys_writer() is called.

**Review by janste63 — CHANGES_REQUESTED** [review]
Missed writer method: do_keys
Missed reader methods: __contains__, __iter__, __len__,  copy
Maybe __iter__ should work on a copy to avoid errors.

File: src/cryptojwt/key_bundle.py (hunk 1 above)
Comment (jschlyter) [reply src/cryptojwt/key_bundle.py]: 670d0bc087c9adfa4ad945f68f80921202d5e138

File: src/cryptojwt/key_bundle.py (hunk 2 above)
Comment (jschlyter) [reply src/cryptojwt/key_bundle.py]: 670d0bc087c9adfa4ad945f68f80921202d5e138

**Comment by jschlyter** [comment]
> Missed writer method: do_keys

Not needed as `update` is wrapped. If one wraps `do_keys` as well you have a conflict.

> Missed reader methods: **contains**, **iter**, **len**, copy

Fixed via caead3fcff17264b6fbe8fd9d282788c6d42ed09

> Maybe **iter** should work on a copy to avoid errors.

670d0bc087c9adfa4ad945f68f80921202d5e138

**Comment by janste63** [comment]
> > Missed writer method: do_keys
> 
> Not needed as `update` is wrapped. If one wraps `do_keys` as well you have a conflict.
> 

Well since the name doesn't start with "_" it suggest that it can be called from the outside. Suggestion, make it private.

The method load() calls do_keys() with no locking.

**Review by janste63 — APPROVED** [review]

**Pull request merged**